src/components/elements/upload/Uploader.tsx

```diff
@@ -0,0 +1,385 @@
+import { ensureArray } from '@/utils/arrays';
+
+const DEFAULT_MAX_BYTES = 3000000;
+const IMAGE_FILE_TYPES = ['.png', '.jpg', '.jpeg', '.gif'];
+
+const getFileTypesFromAccept = (accept: Accept) => {
+  let arr = sortBy(
+    compact(uniq(flatten(Object.values(accept)))).map((e) =>
+      e.toUpperCase().replace(/\.(.*)/, '$1')
+    )
+  );
+
+  if (arr.length === 2) arr = [arr.join(' or ')];
+  if (arr.length > 2) arr = [...arr.slice(0, -2), arr.slice(-2).join(' or ')];
+
+  return arr.join(', ');
+};
+
+const getReadableSize = (maxSize: number) =>
+  `${(maxSize / 1000000).toFixed(1)}MB`;
+
+export type SingleUploaderProps = {
+  multiple?: false;
+  file?: string | FileFieldsFragment;
+  onChange?: (file: string | FileFieldsFragment | undefined) => void;
+  onUpload?: (
+    upload: DirectUpload | undefined,
+    file: File | undefined
+  ) => any | Promise<any>;
+};
+
+export type MultipleUploaderProps = {
+  multiple: true;
+  files?: (string | FileFieldsFragment)[];
+  onChange?: (files: (string | FileFieldsFragment)[]) => void;
+  onUpload?: (uploads: DirectUpload[], files: File[]) => any | Promise<any>;
+};
+
+export type UploaderProps = {
+  accept?: Accept;
+  image?: boolean;
+  maxSize?: number;
+  id: string;
+  ariaLabel?: string | null;
+} & (SingleUploaderProps | MultipleUploaderProps);
+
+const Uploader = ({
+  id,
+  onChange,
+  onUpload,
+  accept: acceptProp,
+  image: isImage = false,
+  maxSize = DEFAULT_MAX_BYTES,
+  multiple,
+  ariaLabel,
+  ...rest
+}: UploaderProps) => {
+  // The uploader accepts a `files` argument which can contain either:
+  // - a STRING which points at a blob ID of a file that has been uploaded within this session, or
+  // - a FileFieldsFragment record which points at a file record in our database, uploaded during a previous session.
+  // `existingFiles` filters the list to only those files that were uploaded some previous time, so we can render them.
+  // The files uploaded during this session, we render this component's internal state, `currentFiles`.
+  const existingFiles: FileFieldsFragment[] = useMemo(() => {
+    let filesArr;
+    if (multiple) {
+      const { files } = rest as MultipleUploaderProps;
+      filesArr = ensureArray(files);
+    } else {
+      const { file } = rest as SingleUploaderProps;
+      filesArr = ensureArray(file);
+    }
+    return filesArr.filter(
+      (f) => typeof f !== 'string'
+    ) as FileFieldsFragment[];
+  }, [rest, multiple]);
+
+  const [currentFiles, setCurrentFiles] = useState<File[]>([]);
+  const [currentUploads, setCurrentUploads] = useState<DirectUpload[]>([]);
+  const [loading, setLoading] = useState<boolean>(false);
+  const [errors, setErrors] = useState<string[]>([]);
+
+  const [uploadFile] = useDirectUpload();
+
+  const uniqueNameValidator = useCallback(
+    (file: File) => {
+      const existingNames = [
+        ...existingFiles.map((f) => (f as FileFieldsFragment).name),
```

Comment:
Do we need to support both `File` and `FileFieldsFragment` as types that can be uploaded? it seems like that is causing some complication with types and needing to cast in some places. I think it could be safe to say that we expect parents of this component to be dealing with `FileFieldsFragment` fragments?

Reply:
Hm, `existingFiles` should always be `FileFieldsFragment` and `currentFiles` should always be type `File`. The `files` arg can contain both `FileFieldsFragment`s and string blob IDs, which are filtered out of `existingFiles` on line 94 (because Uploader is a controlled component when used by DynamicField).

This particular cast is unnecessary thanks to the cast on line 95, so I'll take it out. But let's discuss more since I agree it is confusing, maybe we can look at together to see if it can be improved.

Reply:
I added some more comments to hopefully make it a bit easier to reason through!